test/index.js

```diff
@@ -1206,6 +1206,37 @@ test('elem', function() {
   eq(Z.elem(0, Nothing), false);
 });
 
+test('foldMap', function() {
+  eq(Z.foldMap.length, 3);
+  eq(Z.foldMap.name, 'foldMap');
+
+  // Monoid instance for functions of type a -> a corresponding
+  // to reverse function composition
+  function DualEndo(f) {
+    if (!(this instanceof DualEndo)) return new DualEndo(f);
+    this.runEndo = f;
+  }
+  DualEndo['fantasy-land/empty'] = function() { return DualEndo(x => x); };
+  DualEndo.prototype['fantasy-land/concat'] = function(g) { return DualEndo(a => g.runEndo(this.runEndo(a))); };
+
+  // Derive reduce (foldl) from foldMap
+  function reduce(f, z, x) {
+    function mmap(a) { return DualEndo(function(b) { return f(b, a); }); }
+    var finalEndo = Z.foldMap(DualEndo, mmap, x);
+    return finalEndo.runEndo(z);
+  }
+
+  // Test derived reduce behaves identically to Z.reduce
+  eq(reduce(Z.concat, 'x', []), 'x');
```

Comment:
How about actually comparing the two results? For example:

```diff
-  eq(reduce(Z.concat, 'x', []), 'x');
+  eq(reduce(Z.concat, 'x', []),
+     Z.reduce(Z.concat, 'x', []));
```

We could even define a function to remove the duplication:

```javascript
function compare(f, x, foldable) {
  eq(reduce(f, x, foldable), Z.reduce(f, x, foldable));
}
```

Reply:
I was thinking of doing `function testReduce(testName, reduce) { ... }`, and passing in both the original and `foldMap` derived `reduce`.

Reply:
Regarding directly comparing them to each other, that could work, but we'd still need to independently validate that they're both producing something correct (rather than both of them producing identical wrong results). If we just check that both of them are producing the same (correct) values over a domain of interest, then transitively we get for free that they're also producing identical values.